File: C/206. Reverse Linked List.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

/* #1. Do it iteratively. */
struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode *prev, *cur, *tmp;

    prev = NULL;
    cur = head;
    while (cur != NULL) {
        tmp = cur->next;
        cur->next = prev;
        prev = cur;
        cur = tmp;
    }

    return prev;
}

/* #2. Do it recursively. */
struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode *p;

    if (head == NULL || head->next == NULL) {
        return head;
    } else {
        p = reverseList(head->next);
        head->next->next = head;
        head->next = NULL;
        return p;
    }
}

File: Python3/206. Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        prev, cur = None, head

        while cur:
            prev, cur.next, cur = cur, prev, cur.next
            
        return prev
